working/0678 - partial classes/feature-brainstorm.md

```diff
@@ -0,0 +1,199 @@
+# Partial classes
+
+Author: kevmoo@google.com
+
+Proposed solution to [partial classes (#252)](https://github.com/dart-lang/language/issues/252).
+Discussion about this proposal should go in [Issue #678](https://github.com/dart-lang/language/issues/678).
+
+## Motivation
+
+A critical use of code generation is adding boilerplate to hand-written classes.
+
+An ideal solution would provide:
+
+- Strict separation between human- and computer-generated code.
+- Allow computer-generated code to add members (functions, fields, properties,
+  constructors, etc) to human-generated code directly – without requiring
+  subclasses, mix-ins, or manually connecting generated private members with user-created public members.
+- Allow user-created "stubs" to be added filled in by generated code. This
+  a number of static-analysis and tooling scenarios to work without the code
+  being generated first.
+
+## json_serializable before/after example
+
+Here's a motivating example with a speculative future implementation for
+[json_serializable](https://pub.dev/packages/json_serializable).
+
+### Before (current behavior)
+
+```dart
+// before: User-generated code
+import 'package:json_annotation/json_annotation.dart';
+
+part 'example.g.dart';
+
+@JsonSerializable(nullable: false)
+class Person {
+  final String firstName;
+  final String lastName;
+  final DateTime dateOfBirth;
+  Person({this.firstName, this.lastName, this.dateOfBirth});
+
+  // User needs to connect machine-generated factory. Cannot be a constructor.
+  factory Person.fromJson(Map<String, dynamic> json) => _$PersonFromJson(json);
+
+  // User needs to connect machine-generated function.
+  Map<String, dynamic> toJson() => _$PersonToJson(this);
+}
+```
+
+```dart
+// before: machine-generated
+part of 'example.dart';
+
+// Machine-generated code dirties up the private namespace of the user library
+// with specially named private functions
+Person _$PersonFromJson(Map<String, dynamic> json) {
+  return Person(
+    firstName: json['firstName'] as String,
+    lastName: json['lastName'] as String,
+    dateOfBirth: DateTime.parse(json['dateOfBirth'] as String),
+  );
+}
+
+Map<String, dynamic> _$PersonToJson(Person instance) => <String, dynamic>{
+      'firstName': instance.firstName,
+      'lastName': instance.lastName,
+      'dateOfBirth': instance.dateOfBirth.toIso8601String(),
+    };
+```
+
+### After (possible future)
+
+```dart
+// after: User-generated code
+import 'package:json_annotation/json_annotation.dart';
+
+part 'example.g.dart';
+
+@JsonSerializable(nullable: false)
+class Person {
+  final String firstName;
+  final String lastName;
+  final DateTime dateOfBirth;
+  Person({this.firstName, this.lastName, this.dateOfBirth});
+
+  // No special wiring needed!
```

Comment:
When looking at this class as a user how would I know the toJson/fromJson exist? The annotation is kind of a hint but annotations aren't required here.

You would have to click through to the part file to see these extra memebers and there isn't any indication that there are additional members.

This is less boilerplate though for the user :man_shrugging:

Reply:
ack